expr_checker.py

```diff
@@ -0,0 +1,128 @@
+        self.generic_visit(node)
+
+        if not self.fncall:
+            raise Exception("safe_eval didn't permit you to call any functions")
+
+        if isinstance(node.func, ast.Attribute):
+            return ast.Call(
+                func=ast.Name(self.getattr, ast.Load()),
+                args=[node.func.value, ast.Constant(node.func.attr), node],
+                keywords=[]
+            )
+
+        else:
+            self.generic_visit(node.func)
+
+        return ast.Call(
+            func=ast.Name(self.check_fn, ctx=ast.Load()),
+            args=[node.func, ast.Name(
+                self.check_type_fn, ctx=ast.Load())] + node.args,
+            keywords=node.keywords
+        )
+
+    def visit_FunctionDef(self, node):
+        self.generic_visit(node)
+        if node.name in self.reserved_name:
+            raise NameError(f"safe_eval: {node.name} is a reserved name")
+
+        return node
+
+    def visit_Attribute(self, node):
+        self.generic_visit(node.value)
+
+        if isinstance(node.ctx, ast.Load):
+            subcall = ast.Call(
+                func=ast.Name(self.getattr, ctx=ast.Load()),
+                args=[node.value, ast.Constant(node.attr), node],
+                keywords=[]
+            )
+
+            return ast.Call(
+                func=ast.Name(self.check_type_fn, ctx=ast.Load()),
+                args=[ast.Constant("attribute"), subcall],
+                keywords=[]
+            )
+
+        elif isinstance(node.ctx, ast.Store):
+            raise ValueError("safe_eval: doesn't permit you to store values in attributes")
+
+        elif isinstance(node.ctx, ast.Del):
+            raise ValueError("safe_eval: doesn't permit you to delete attributes")
+
+    def visit_Assign(self, node):
+        self.generic_visit(node.value)
+
+        for target in node.targets:
+            if isinstance(target, ast.Name):
+                if target.id in self.reserved_name:
+                    raise NameError(f"safe_eval: {target.id} is a reserved name")
+            elif isinstance(target, ast.Attribute):
+                raise ValueError("safe_eval: doesn't permit you to modify an attribute")  # FIXME with ast_set_attr
+            elif isinstance(target, ast.Subscript) and isinstance(target.value,
+                                                                  ast.Call) and target.value.func.id in ["globals",
+                                                                                                         "locals"]:
+                raise ValueError("safe_eval: doesn't permit you to modify locals() and globals()")
+            else:
+                raise NotImplementedError(f"{ast.dump(target, indent=4)}")
+
+        return node
+
+    def visit_Subscript(self, node):
+        self.generic_visit(node)
+        return ast.Call(
+            func=ast.Name(self.check_type_fn, ctx=ast.Load()),
+            args=[ast.Constant("constant"), node],
+            keywords=[]
+        )
```

Comment:
looks like I still can bypass this
```python
    def test_std(self):

        file = open('/etc/passwd')

        def test():
            return file, None

        code = """
            a = test()
            b = next(x for x in a)

        """

        exec(expr_checker(cleandoc(code), ast_get_attr, check_type=check_type))

        file.close()
```

I wonder if there's no "more robust" way to check the types